### PR TITLE
Follow up improvements to `Date_Time_Formatter`

### LINKED ISSUE #7866: Make `Data_Formatter.parse` use `Value_Type`
`Data_Formatter.parse` is using raw enso types as argument to determine what type to parse into. This is slightly hacky and comes from old times before we had `Value_Type`.

`Value_Type` will fit here better and will be consistent with other APIs.

This is a very small ticket to update it.

### LINKED ISSUE #7872: Additional improvements to date-time format
The PR #7826 improved date-time format parsing to fix #7461, but it already has exceeded its scope.

This task tracks a few small follow-ups that we want to do but to not grow the mentioned PR even larger, they will be done separately:

- [x] Try migrating custom parse logic in `parseLocalDate` to instead be done at construction of the formatter and rely on `parseDefaulting`.
- [x] Add warnings for M/m mismatch based on surroundings.
- [x] Add a warning when 2 patterns for seconds are detected - to prevent confusion.
	- In Java one could write `HH:mm:ss.S` to have fractions. But our format is mostly case insensitive and uses `f` for fractions instead. So if we see `ss.S` we report the warning and suggest to use `f` instead. Correct format is `HH:mm:ss.f`.
- [x] Add a warning for 24h used with am/pm.
- [x] Add a warning for 12h without am/pm BUT allow it to parse (defaulting to AM).
- [x] Add a warning when parsing `MM-dd` (`yyyy` defaults to current year then). No warning on format!

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

- Closes #7872 
- Also closes #7866 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Review by GregoryTravis — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Internal/Time/Format/Analyzer.enso [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+    ## PRIVATE
+    has_12h : Boolean
+    has_12h self =
+        self.flattened.any node-> case node of
+            Time_Patterns.Hour _ is24h -> is24h.not
+            _ -> False
+
+    ## PRIVATE
+    has_ampm : Boolean
+    has_ampm self =
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Internal/Time/Format/Analyzer.enso]: Should we use `am_pm` throughout?
```suggestion
    has_am_pm : Boolean
    has_am_pm self =
```

**Review by jdunkerley — APPROVED** [review]
one slight question but looks good.

**Comment by radeusgd** [comment]
Added a benchmark (to Exploratory Benchmarks, so it will not be run by CI by default, I don't think it's important enough to run it, but IMO worth keeping in repo).

<details>
<summary>Raw results</summary>

```
Found 4 cases to execute
Benchmarking 'Format_Vector_Of_Dates.Naive' with configuration: [warmup={2 iterations, 5 seconds each}, measurement={2 iterations, 5 seconds each}]
Warmup duration:    11752.7002 ms
Warmup invocations: 3
Warmup avg time:    3910.385 ms
Measurement duration:    12369.5478 ms
Measurement invocations: 5
Measurement avg time:    2473.712 ms
Benchmark 'Format_Vector_Of_Dates.Naive' finished in 24129.273 ms
Benchmarking 'Format_Vector_Of_Dates.Prepared_Formatter' with configuration: [warmup={2 iterations, 5 seconds each}, measurement={2 iterations, 5 seconds each}]
Warmup duration:    10190.8434 ms
Warmup invocations: 1681
Warmup avg time:    5.949 ms
Measurement duration:    10242.1637 ms
Measurement invocations: 2431
Measurement avg time:    4.115 ms
Benchmark 'Format_Vector_Of_Dates.Prepared_Formatter' finished in 20434.452 ms
Benchmarking 'Format_Vector_Of_Dates.Java_Formatter_recreated_on_each' with configuration: [warmup={2 iterations, 5 seconds each}, measurement={2 iterations, 5 seconds each}]
Warmup duration:    10040.4148 ms
Warmup invocations: 793
Warmup avg time:    12.621 ms
Measurement duration:    10045.1645 ms
Measurement invocations: 930
Measurement avg time:    10.761 ms
Benchmark 'Format_Vector_Of_Dates.Java_Formatter_recreated_on_each' finished in 20086.788 ms
Benchmarking 'Format_Vector_Of_Dates.Column_Format' with configuration: [warmup={2 iterations, 5 seconds each}, measurement={2 iterations, 5 seconds each}]
Warmup duration:    10026.1589 ms
Warmup invocations: 549
Warmup avg time:    18.221 ms
Measurement duration:    10044.7375 ms
Measurement invocations: 637
Measurement avg time:    15.713 ms
Benchmark 'Format_Vector_Of_Dates.Column_Format' finished in 20072.428 ms
```

</details>

Re-creating the formatter at each invocation (default if you call something like `vector.map _.format "str"`) is about **600x slower** (!) than using a prepared formatter instance.

In comparison, using `Column.format` (which creates a formatter under the hood), is only ~4-5 times slower.

We should indeed try adding an LRU cache for the formats.

<details>
<summary>Results with shorter warmup (more CI ready)</summary>

```
Found 4 cases to execute
Benchmarking 'Format_Vector_Of_Dates.Naive' with configuration: [warmup={1 iterations, 4 seconds each}, measurement={2 iterations, 4 seconds each}]
Warmup duration:    6927.1965 ms
Warmup invocations: 1
Warmup avg time:    6910.172 ms
Measurement duration:    8217.4258 ms
Measurement invocations: 3
Measurement avg time:    2738.048 ms
Benchmark 'Format_Vector_Of_Dates.Naive' finished in 15151.292 ms
Benchmarking 'Format_Vector_Of_Dates.Prepared_Formatter' with configuration: [warmup={1 iterations, 4 seconds each}, measurement={2 iterations, 4 seconds each}]
Warmup duration:    4048.7626 ms
Warmup invocations: 358
Warmup avg time:    11.188 ms
Measurement duration:    8127.921 ms
Measurement invocations: 969
Measurement avg time:    8.264 ms
Benchmark 'Format_Vector_Of_Dates.Prepared_Formatter' finished in 12178.561 ms
Benchmarking 'Format_Vector_Of_Dates.Java_Formatter_recreated_on_each' with configuration: [warmup={1 iterations, 4 seconds each}, measurement={2 iterations, 4 seconds each}]
Warmup duration:    4008.1436 ms
Warmup invocations: 114
Warmup avg time:    35.117 ms
Measurement duration:    8023.6116 ms
Measurement invocations: 562
Measurement avg time:    14.244 ms
Benchmark 'Format_Vector_Of_Dates.Java_Formatter_recreated_on_each' finished in 12033.032 ms
Benchmarking 'Format_Vector_Of_Dates.Column_Format' with configuration: [warmup={1 iterations, 4 seconds each}, measurement={2 iterations, 4 seconds each}]
Warmup duration:    4016.2714 ms
Warmup invocations: 172
Warmup avg time:    23.333 ms
Measurement duration:    8015.4114 ms
Measurement invocations: 413
Measurement avg time:    19.373 ms
Benchmark 'Format_Vector_Of_Dates.Column_Format' finished in 12033.088 ms
```

</details>

File: distribution/lib/Standard/Base/0.0.0-dev/src/Internal/Time/Format/Analyzer.enso (hunk 1 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Internal/Time/Format/Analyzer.enso]: It's just internal stuff, but yeah I guess that will be better - I will amend it.

**Comment by radeusgd** [comment]
Benchmarks after implementing the cache

<details>
<summary>Raw results</summary>

```
Found 4 cases to execute
Benchmarking 'Format_Vector_Of_Dates.Naive' with configuration: [warmup={1 iterations, 4 seconds each}, measurement={2 iterations, 4 seconds each}]
Warmup duration:    4051.5825 ms
Warmup invocations: 372
Warmup avg time:    10.753 ms
Measurement duration:    8104.8076 ms
Measurement invocations: 1285
Measurement avg time:    6.229 ms
Benchmark 'Format_Vector_Of_Dates.Naive' finished in 12166.056 ms
Benchmarking 'Format_Vector_Of_Dates.Prepared_Formatter' with configuration: [warmup={1 iterations, 4 seconds each}, measurement={2 iterations, 4 seconds each}]
Warmup duration:    4051.4442 ms
Warmup invocations: 831
Warmup avg time:    4.814 ms
Measurement duration:    8175.2936 ms
Measurement invocations: 2022
Measurement avg time:    3.957 ms
Benchmark 'Format_Vector_Of_Dates.Prepared_Formatter' finished in 12228.495 ms
Benchmarking 'Format_Vector_Of_Dates.Java_Formatter_recreated_on_each' with configuration: [warmup={1 iterations, 4 seconds each}, measurement={2 iterations, 4 seconds each}]
Warmup duration:    4008.8002 ms
Warmup invocations: 259
Warmup avg time:    15.449 ms
Measurement duration:    8026.8088 ms
Measurement invocations: 682
Measurement avg time:    11.734 ms
Benchmark 'Format_Vector_Of_Dates.Java_Formatter_recreated_on_each' finished in 12036.888 ms
Benchmarking 'Format_Vector_Of_Dates.Column_Format' with configuration: [warmup={1 iterations, 4 seconds each}, measurement={2 iterations, 4 seconds each}]
Warmup duration:    4013.7933 ms
Warmup invocations: 214
Warmup avg time:    18.735 ms
Measurement duration:    8061.3893 ms
Measurement invocations: 832
Measurement avg time:    9.617 ms
Benchmark 'Format_Vector_Of_Dates.Column_Format' finished in 12076.766 ms
```

</details>

| Approach | Avg Measurement Time |
| --- | ---: |
| Naïve (without cache) | 2738.048 ms |
| Naïve (with cache) | 6.229 ms |
| Formatter precomputed | 3.957 ms |
| Java `ofPattern` re-created on each call | 11.734 ms |
| `Column.format` | 9.617 ms |

Adding the cache was definitely worth it - on this example we are about 400x faster than before!

After implementing the cache all timings are on par between ~4-11ms.

**Pull request merged**